File: modules/multiplexer/transport.go
package multiplexer

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// IPC message types exchanged over the socket.
const (
	IPCTask      = "task"
	IPCResult    = "result"
	IPCEvent     = "event"
	IPCKeepalive = "keepalive"
)

// IPCMessage is the envelope for all inter-pane communication.
type IPCMessage struct {
	Type    string          `json:"type"`
	ID      string          `json:"id"`
	Payload json.RawMessage `json:"payload"`
}

// TaskPayload is sent parent -> child to assign work.
type TaskPayload struct {
	Task      string            `json:"task"`
	AgentID   string            `json:"agent_id"`
	SessionID string            `json:"session_id"`
	Tools     []string          `json:"tools"`
	Env       map[string]string `json:"env,omitempty"`
}

// ResultPayload is sent child -> parent when the task finishes.
type ResultPayload struct {
	Success bool   `json:"success"`
	Content string `json:"content"`
	Error   string `json:"error,omitempty"`
}

// EventPayload is sent child -> parent for streaming updates.
type EventPayload struct {
	EventType string `json:"event_type"`
	Data      string `json:"data"`
}

// SocketTransport implements bidirectional IPC over a Unix domain socket.
// It satisfies a line-delimited JSON protocol compatible with the mcp.Transport pattern.
// Send is safe for concurrent use; Receive is not.
type SocketTransport struct {
	conn    net.Conn
	reader  *bufio.Reader
	writer  *bufio.Writer
	writeMu sync.Mutex
}

// NewSocketTransport wraps an existing net.Conn.
func NewSocketTransport(conn net.Conn) *SocketTransport {
	return &SocketTransport{
		conn:   conn,
		reader: bufio.NewReader(conn),
		writer: bufio.NewWriter(conn),
	}
}

// Send writes a JSON-encoded message as a single newline-terminated line.
// Safe for concurrent use.
func (t *SocketTransport) Send(msg IPCMessage) error {
	data, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("marshal IPC message: %w", err)
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	if _, err := t.writer.Write(data); err != nil {
		return err
	}
	if err := t.writer.WriteByte('\n'); err != nil {
		return err
	}
	return t.writer.Flush()
}

// Receive reads the next newline-terminated JSON message.
func (t *SocketTransport) Receive() (IPCMessage, error) {
	line, err := t.reader.ReadBytes('\n')
	if err != nil {
		return IPCMessage{}, err
	}
	var msg IPCMessage
	if err := json.Unmarshal(line[:len(line)-1], &msg); err != nil {
		return IPCMessage{}, fmt.Errorf("unmarshal IPC message: %w", err)
	}
	return msg, nil
}

// Close closes the underlying connection.
func (t *SocketTransport) Close() error {
	return t.conn.Close()
}

// CreateListener creates a Unix domain socket listener at the given path.
// Parent side calls this to wait for the child process to connect.
func CreateListener(socketPath string) (net.Listener, error) {
	if err := os.MkdirAll(filepath.Dir(socketPath), 0755); err != nil {
		return nil, fmt.Errorf("create socket dir: %w", err)
	}
	// Remove stale socket if present.
	os.Remove(socketPath)

	ln, err := net.Listen("unix", socketPath)
	if err != nil {
		return nil, fmt.Errorf("listen on %s: %w", socketPath, err)
	}
	return ln, nil
}

// WaitForConnection accepts a single connection with a timeout.
func WaitForConnection(ln net.Listener, timeout time.Duration) (net.Conn, error) {
	type result struct {
		conn net.Conn
		err  error
	}
	ch := make(chan result, 1)
	go func() {
		conn, err := ln.Accept()
		ch <- result{conn, err}
	}()

	select {
	case r := <-ch:
		return r.conn, r.err
	case <-time.After(timeout):
		return nil, fmt.Errorf("timeout waiting for pane connection after %s", timeout)
	}
}

// ConnectSocket connects to a Unix domain socket at the given path.
// Child side calls this to establish IPC with the parent.
func ConnectSocket(socketPath string, timeout time.Duration) (net.Conn, error) {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		conn, err := net.Dial("unix", socketPath)
		if err == nil {
			return conn, nil
		}
		time.Sleep(50 * time.Millisecond)
	}
	return nil, fmt.Errorf("timeout connecting to socket %s after %s", socketPath, timeout)
}

// SocketPath returns the standard socket path for a given pane ID.
func SocketPath(socketDir string, paneID string) string {
	return filepath.Join(socketDir, paneID+".sock")
}

// CleanupSocket removes the socket file and its directory if empty.
func CleanupSocket(socketPath string) {
	os.Remove(socketPath)
	dir := filepath.Dir(socketPath)
	// Try to remove dir; ignore error if not empty.
	os.Remove(dir)
}

File: packages/multiplexer/src/transport.ts
import { createConnection, createServer, type Socket } from "node:net";
import { mkdir, rm } from "node:fs/promises";
import { join, dirname } from "node:path";
import { newIOError } from "@orangecoding/core";

// ---------------------------------------------------------------------------
// IPC message types
// ---------------------------------------------------------------------------

/** IPC message type constants. */
export const IPCMessageType = {
  Task: "task",
  Result: "result",
  Event: "event",
  Keepalive: "keepalive",
} as const;
export type IPCMessageType = (typeof IPCMessageType)[keyof typeof IPCMessageType];

/** IPCMessage is the envelope for all inter-pane communication. */
export interface IPCMessage {
  type: IPCMessageType;
  id: string;
  payload: unknown;
}

/** TaskPayload is sent parent -> child to assign work. */
export interface TaskPayload {
  task: string;
  agentId: string;
  sessionId: string;
  tools: string[];
  env?: Record<string, string>;
}

/** ResultPayload is sent child -> parent when the task finishes. */
export interface ResultPayload {
  success: boolean;
  content: string;
  error?: string;
}

/** EventPayload is sent child -> parent for streaming updates. */
export interface EventPayload {
  eventType: string;
  data: string;
}

// ---------------------------------------------------------------------------
// SocketTransport
// ---------------------------------------------------------------------------

/**
 * SocketTransport implements bidirectional IPC over a Unix domain socket
 * using a line-delimited JSON protocol.
 */
export class SocketTransport {
  private buffer = "";

  constructor(private conn: Socket) {}

  /**
   * Send writes a JSON-encoded message as a single newline-terminated line.
   */
  async send(msg: IPCMessage): Promise<void> {
    const data = JSON.stringify(msg) + "\n";
    return new Promise((resolve, reject) => {
      this.conn.write(data, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  /**
   * Receive reads the next newline-terminated JSON message.
   */
  async receive(): Promise<IPCMessage> {
    return new Promise((resolve, reject) => {
      const onData = (chunk: Buffer | string) => {
        this.buffer += chunk.toString();
        const idx = this.buffer.indexOf("\n");
        if (idx !== -1) {
          const line = this.buffer.substring(0, idx);
          this.buffer = this.buffer.substring(idx + 1);
          this.conn.removeListener("data", onData);
          this.conn.removeListener("error", onError);
          this.conn.removeListener("end", onEnd);
          try {
            resolve(JSON.parse(line));
          } catch (err) {
            reject(newIOError(`unmarshal IPC message: ${(err as Error).message}`));
          }
        }
      };

      const onError = (err: Error) => {
        this.conn.removeListener("data", onData);
        this.conn.removeListener("end", onEnd);
        reject(err);
      };

      const onEnd = () => {
        this.conn.removeListener("data", onData);
        this.conn.removeListener("error", onError);
        reject(new Error("socket closed"));
      };

      this.conn.on("data", onData);
      this.conn.on("error", onError);
      this.conn.on("end", onEnd);
    });
  }

  /** Close closes the underlying connection. */
  async close(): Promise<void> {
    this.conn.destroy();
  }
}

// ---------------------------------------------------------------------------
// Socket helpers
// ---------------------------------------------------------------------------

/**
 * CreateListener creates a Unix domain socket listener at the given path.
 * Parent side calls this to wait for the child process to connect.
 */
export async function createListener(socketPath: string) {
  try {
    await mkdir(dirname(socketPath), { recursive: true });
  } catch (err) {
    throw newIOError(`create socket dir: ${(err as Error).message}`);
  }

  // Remove stale socket if present.
  try { await rm(socketPath); } catch { /* ignore */ }

  const server = createServer();
  return new Promise<typeof server>((resolve, reject) => {
    server.listen(socketPath, () => resolve(server));
    server.on("error", reject);
  });
}

/**
 * WaitForConnection accepts a single connection with a timeout.
 */
export async function waitForConnection(
  server: import("node:net").Server,
  timeoutMs: number,
): Promise<Socket> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error(`timeout waiting for pane connection after ${timeoutMs}ms`));
    }, timeoutMs);

    server.once("connection", (socket: Socket) => {
      clearTimeout(timer);
      resolve(socket);
    });
  });
}

/**
 * ConnectSocket connects to a Unix domain socket at the given path.
 * Child side calls this to establish IPC with the parent.
 */
export async function connectSocket(socketPath: string, timeoutMs: number): Promise<Socket> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    try {
      return await new Promise<Socket>((resolve, reject) => {
        const sock = createConnection(socketPath, () => resolve(sock));
        sock.on("error", reject);
      });
    } catch {
      await new Promise((r) => setTimeout(r, 50));
    }
  }
  throw new Error(`timeout connecting to socket ${socketPath} after ${timeoutMs}ms`);
}

/**
 * SocketPath returns the standard socket path for a given pane ID.
 */
export function socketPath(socketDir: string, paneID: string): string {
  return join(socketDir, `${paneID}.sock`);
}

/**
 * CleanupSocket removes the socket file and its directory if empty.
 */
export async function cleanupSocket(path: string): Promise<void> {
  try { await rm(path); } catch { /* ignore */ }
  try { await rm(dirname(path)); } catch { /* ignore if not empty */ }
}
